### DISCUSSION TITLE
Interrupts for Serial port incoming data

### DISCUSSION THREAD

**Discussion by ardlc-1212** [earlephilhower/arduino-pico#1596] [Q&A]
Hi, I have been trying to set up a code that uses interrupts to give me a notification when there is a byte to read in the Serial input buffer. I wasn't able to find any documentation on it but theoretically speaking, page 316 of rp2040 datasheet "3.2.6. IRQ Flags", if I'm not wrong, mentions that we can create system level interrupts from the state machines which can be used to inform the processor of a incoming byte completion or even completion of x amount of bits/bytes. Is there any implementation of this? Help would be greatly appreciated also thanks for giving me your time <3

**Comment by earlephilhower** [comment]
By `Serial` you do mean one of the UARTs, yes (uart0/Serial1 or uart1/Serial2)?  `Serial` is the USB port and is very different.

If for some reason you want to re-implement the IRQ handler for the UART ports, I suggest looking at https://github.com/earlephilhower/arduino-pico/blob/master/cores/rp2040/SerialUART.cpp which has both polled and IRQ driven UART control.

**Comment by ardlc-1212** [comment]
Sorry let me clear it for you:
1. I mean UART, specifically Serial2
2. What I mean by Interrupt isn't polled or IRQ driven it should be IRQ driven there bc of the near 500,000bps rate but rather I mean that for every byte that arrives I want a subroutine to be activated (like gpio interrupt) i dont know if that makes sense I can make a diagram too if it's still not clear. 

Thanks, Arda

**Reply by earlephilhower** [reply]
You need to write your own code to do it., but that's simple.  Just set the UARTIFLS.RXIFLSEL to 0b000 (IRQ on >=1 ready) and enable the appropriate interrupts in the UART and hook it using the SDK APIs.

That said, getting 50K interrupts/second is probably the wrong way of handling whatever you're trying to do.  Might work, but there is overhead on each IRQ enter and leave so don't expect much work in the main code to actually get done.

**Reply by ardlc-1212** [reply]
That isnt a problem I have a really low amount of data income but each transfer has to be fast and I didnt wanted to keep polling for Serial.available() on my Handler, I am using FreeRTOS so it would have to constantly switch to this handler check it and go back which is really inefficient. Thanks for the guidance earlephilhower <3

**Reply by ardlc-1212** [reply]
Hey sorry I have another problem, I am getting an error code that doesnt really explain anything

`
.pio\libdeps\pico\SerialUARTModified\SerialUARTModified.cpp:462:62: error: 'PIN_Serial1_TX' was not declared in this scope
  462 | SerialUARTModified Serial1Modified(__Serial1Modified_DEVICE, PIN_Serial1_TX, PIN_Serial1_RX);

.pio\libdeps\pico\SerialUARTModified\SerialUARTModified.cpp:462:78: error: 'PIN_Serial1_RX' was not declared in this scope
  462 | SerialUARTModified Serial1Modified(__Serial1Modified_DEVICE, PIN_Serial1_TX, PIN_Serial1_RX);

.pio\libdeps\pico\SerialUARTModified\SerialUARTModified.cpp:468:62: error: 'PIN_Serial_TX' was not declared in this scope
  468 | SerialUARTModified Serial2Modified(__Serial2Modified_DEVICE, PIN_Serial_TX, PIN_Serial_RX);

.pio\libdeps\pico\SerialUARTModified\SerialUARTModified.cpp:468:77: error: 'PIN_Serial_RX' was not declared in this scope
  468 | SerialUARTModified Serial2Modified(__Serial2Modified_DEVICE, PIN_Serial_TX, PIN_Serial_RX);

*** [.pio\build\pico\libbfb\SerialUARTModified\SerialUARTModified.cpp.o] Error 1
`

it really doesn't make sense to me doesnt these have to be pre-defined?

**Reply by earlephilhower** [reply]
Try `PIN_SERIAL1_TX` or `PIN_SERIAL2_TX`.  All caps for defines, and there is no `PIN_SERIAL_TX` w/o a # because `Serial` does not have RX or TX pins.

**Reply by ardlc-1212** [reply]
In the Serial1 I did that as you can see from the previous comment but that still gave an error. Just in case, I tried to add them both as Serial# but no help still the same errors. 

Also I dont know if it would help but I am including this as a library and not modifying from the root directory maybe that could be the problem but, u know, I dont wanna mess with there this is a more robust approach in my opinion. 

Also thanks again for answering all my dumb questions <3

**Reply by earlephilhower** [reply]
Sorry, I don't know how you're not getting them defined.  Are you including `<Arduino.h>`?  They are set for all boards.  For the RPi Pico they're here:
https://github.com/earlephilhower/arduino-pico/blob/1dc0872818cab9631e9577acf8135c8c417612fc/variants/rpipico/pins_arduino.h#L10-L15

**Reply by ardlc-1212** [reply]
Oh, nevermind I just found the issue it was that when I was making them all caps for defines I had a small typo that made them PIM_SERIAL#_TX instead of pin I fixed that and now its working. Thank you for the help

Wish you the best, 
     Arda
